graphql-protocol/src/main/java/com/linecorp/armeria/server/graphql/protocol/MultipartVariableMapper.java

```diff
@@ -0,0 +1,148 @@
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.multipart.MultipartFile;
+
+/**
+ * Mapper for handling populating the query variables with the files specified by object paths
+ * in the multi-part request.
+ */
+final class MultipartVariableMapper {
+
+   // Forked form https://github.com/Netflix/dgs-framework/blob/12cc1931e3c410e342134b7c0a8401f2c5a4aca6/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/utils/MultipartVariableMapper.kt
+
+   private static final int INVALID_INT = -1;
+   private static final Splitter SPLITTER = Splitter.on(".");
+   private static final Mapper<Map<String, Object>> MAP_MAPPER = new Mapper<Map<String, Object>>() {
+      @Override
+      public Object set(Map<String, Object> location, String target, MultipartFile multipartFile) {
+         return location.put(target, multipartFile);
+      }
+
+      @Override
+      public Object recurse(Map<String, Object> location, String target) {
+         return location.get(target);
+      }
+   };
+
+   private static final Mapper<List<Object>> LIST_MAPPER = new Mapper<List<Object>>() {
+      @Override
+      public Object set(List<Object> location, String target, MultipartFile multipartFile) {
+         final int value = parseUnsignedInt(target);
+         if (value == INVALID_INT) {
+            return location;
+         }
+         if (value >= location.size()) {
+            return location;
+         }
+         return location.set(value, multipartFile);
+      }
+
+      @Override
+      public Object recurse(List<Object> location, String target) {
+         final int value = parseUnsignedInt(target);
+         if (value == INVALID_INT) {
+            return null;
+         }
+         if (value >= location.size()) {
+            return null;
+         }
+         return location.get(value);
+      }
+   };
+
+   static void mapVariable(String objectPath, Map<String, Object> variables, MultipartFile multipartFile) {
+      final List<String> segments = ImmutableList.copyOf(SPLITTER.split(objectPath));
+      if (segments.size() < 2) {
+         throw new IllegalArgumentException("Invalid object-path: " + objectPath);
+      }
+      if (!"variables".equals(segments.get(0))) {
+         throw new IllegalArgumentException("Can only map into variables: " + objectPath);
+      }
+
+      Object currentLocation = variables;
+      for (int i = 1; i < segments.size(); i++) {
+         final String segmentName = segments.get(i);
+         if (i == segments.size() - 1) {
+            if (currentLocation instanceof Map) {
+               if (MAP_MAPPER.set((Map<String, Object>) currentLocation, segmentName, multipartFile) != null) {
+                  throw new IllegalArgumentException("Expected null value when mapping: " + objectPath);
+               }
+            } else {
+               if (LIST_MAPPER.set((List<Object>) currentLocation, segmentName, multipartFile) != null) {
```

Comment:
> currentLocation may not necessary be a list at this point, resulting in status code 500 instead of 400 which seems more appropriate.

Wouldn't 400 be more correct because there could be behaviors like the client entering an incorrect index?

> What do you think of explicitly checking currentLocation instanceof List before this block?

good idea!